File: clip-encode-decode/overlong_utf8.js
function toOverlongUTF8(hex) {
  const codePoint = parseInt(hex, 16);
  let m = (b) => '%' + b.toString(16).toUpperCase();
  if (codePoint < 0x80) {
    let two_byte = [0xC0 | (codePoint >> 6), 0x80 | (codePoint & 0x3F)]
    let three_byte = [0xE0,0x80,0x80 | (codePoint & 0x3F)]
    return two_byte.map(m).join('')+' / '+three_byte.map(m).join('');
  } else {
    throw new Error("Only works for ASCII characters");
  }
}

console.log(toOverlongUTF8("3e"));

File: clip-encode-decode/overlong_utf8.py
import sys


def char_to_4_byte_utf8(char):
    # Ensure the character is a single character (not a string)
    if len(char) != 1:
        raise ValueError("Input must be a single character")

    def m(b):
        return '%'+format(b,'X')
    # Get the Unicode code point of the character
    code_point = ord(char)

    # Check if the code point requires a 4-byte representation

    # Convert the code point to a 4-byte UTF-8 encoding
    utf8_bytes = bytes([0xF0 | ((code_point >> 18) & 0x07),
                        0x80 | ((code_point >> 12) & 0x3F),
                        0x80 | ((code_point >> 6) & 0x3F),
                        0x80 | (code_point & 0x3F)])

    # Return the 4-byte UTF-8 encoding as bytes
    
    return ''.join(map(m,utf8_bytes))

def to_overlong_utf8(hex):
    code_point = int(hex, 16)

    def m(b):
        return '%' + format(b, 'X')

    if code_point < 0x80:
        two_byte = [0xC0 | (code_point >> 6), 0x80 | (code_point & 0x3F)]
        three_byte = [0xE0, 0x80, 0x80 | (code_point & 0x3F)]
        return ''.join(map(m, two_byte)) + ' / ' + ''.join(map(m, three_byte))
    else:
        raise ValueError("Only works for ASCII characters")

if len(sys.argv) < 2:
    sys.exit(1)


print(to_overlong_utf8(hex(ord(sys.argv[1]))[2:]))

print("\n : "+str(char_to_4_byte_utf8(sys.argv[1])))
